src/lib/ndpi_main.c

```diff
@@ -4249,11 +4249,35 @@ static int ndpi_init_packet_header(struct ndpi_detection_module_struct *ndpi_str
 	u_int16_t backup1, backup2;
 	u_int16_t packet_direction_counter[2];
 
+	const struct ndpi_iphdr *iph;
+	const struct ndpi_ipv6hdr *iphv6;
+	const u_int8_t *payload;
+	u_int64_t current_time_ms;
+	u_int16_t l3_packet_len;
+	u_int16_t l4_packet_len;
+	u_int16_t payload_packet_len;
+	u_int16_t actual_payload_len;
+	u_int8_t l4_protocol;
+
+#define packet_save(a) a = flow->packet.a
+#define packet_restore(a) flow->packet.a = a
+
```

Comment:
Is it feasible to use these macros for the fields we already restore, too? (guessed_protocol_id, num_processed_pkts....)
This way the code should be more uniform